### PR TITLE
Fix edge case issue with extracting variants from less conventional source_locations

### PR DESCRIPTION
**Pull request by rdavid1099**
Problem
=======
I encountered an edge case when trying to use variants. My local development environment has an additional `.` in it (`/example/fake.user/path/to/actionview-component`). The extracting the variant from the `source_location` assumes the path only has `.` in the respective files (`path.split(".").second.split("+")[1]&.to_sym`).

Solution
=======
Using similar patterns to how Rails extracts the variants from views (https://github.com/rails/rails/blob/d2e8b839d2e7236b5f85797f239b01ca2e783024/actionview/lib/action_view/template/resolver.rb#L272), I altered the extraction of the variants in `ActionView::Component::Base#templates`.

_Note_: I did not add any tests for this edge case since, at first, four tests dealing with variants were failing in my local environment. The fix applied has all tests now passing.

### CODE REVIEW

File: lib/action_view/component/base.rb [hunk 1]
```diff
@@ -172,10 +172,12 @@ def identifier
         def templates
           @templates ||=
             (Dir["#{source_location.sub(/#{File.extname(source_location)}$/, '')}.*{#{ActionView::Template.template_handler_extensions.join(',')}}"] - [source_location]).each_with_object([]) do |path, memo|
+              pieces = File.basename(path).split(".")
+
               memo << {
                 path: path,
-                variant: path.split(".").second.split("+")[1]&.to_sym,
-                handler: path.split(".").last
+                variant: pieces.second.split("+")[1]&.to_sym,
```
Comment (juanmanuelramallo) [inline lib/action_view/component/base.rb]: What happens if the path contains several folders with dots in them?

Comment (rdavid1099) [reply lib/action_view/component/base.rb]: 💯 Great idea. I get a few tests in there for this edge case.

File: test/action_view/component/base/unit_test.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+require "test_helper"
+
+class ActionView::Component::Base::UnitTest < Minitest::Test
```
Comment (rdavid1099) [inline test/action_view/component/base/unit_test.rb]: @juanmanuelramallo I added a new unit_test file to isolate the `templates` method and stub out several unique paths. I'm not sure if you want to add unit tests or relocate this test to an existing file. Tell me what you think 😄

Comment (juanmanuelramallo) [reply test/action_view/component/base/unit_test.rb]: This is good. But I feel it would be better to maintain the current approach of testing. What do you think?

That would be about creating new components in `/test/app/components` and then use those in your test cases in `/test/action_view/component_test.rb` so that we see can assert that the output of the rendering is what we expect.

Those new components can have the path names you used in here:
- fake.user/path/to.templates/component/test_component.html+phone.erb
- /_underscore-dash./component/test_component.html+desktop.slim
- /tilda~/component/test_component.html.haml

File: lib/action_view/component/base.rb [hunk 3]
```diff
@@ -169,13 +169,19 @@ def identifier
 
         private
 
+        def matching_views_in_source_location
```
Comment (juanmanuelramallo) [inline lib/action_view/component/base.rb]: If we follow this comment https://github.com/github/actionview-component/pull/123/files#r349892608 we may not need to extract this into a method right? Cause it seems like it was extracted just to be able to stub it in the test no?

File: test/action_view/component/base/unit_test.rb (hunk 2 above)
Comment (rdavid1099) [reply test/action_view/component/base/unit_test.rb]: Sounds good. I'll rework this test and follow the existing testing patterns.

Comment (rdavid1099) [reply test/action_view/component/base/unit_test.rb]: @juanmanuelramallo I'm running into an issue with initializing the component due to Rails conventions.

I have the component and its respective view in the directory path `test/app/components/unconventional.path/_underscore-dash./tilda~/`. When I run the test for this unconventional path:
```ruby
  def test_renders_components_with_less_conventional_paths
    with_variant :phone do
      result = render_inline(UnconventionalPathComponent)

      assert_includes result.text, "Phone"
    end
  end
```
I get a NameError `uninitialized constant ActionView::ComponentTest::UnconventionalPath`. This makes sense since the component is not name spaced using the unconventional directory names. Any ideas on how to get this test to work?

**Comment by joelhawksley** [comment]
@rdavid1099 @juanmanuelramallo any thoughts on how we'd like to proceed with this PR?

File: test/action_view/component/base/unit_test.rb (hunk 2 above)
Comment (juanmanuelramallo) [reply test/action_view/component/base/unit_test.rb]: Sorry for the delay, I lost track of time. 

Maybe we have two options:
1. Use modules to namespace according to the folders (following rails convention)
2. Require the file directly in the spec_helper or in the test file

Anyways I now see that the way you tested it is great! So maybe we can make this go forward as it is, and then we can add some tests following the other approach in a follow up PR

File: lib/action_view/component/base.rb (hunk 1 above)
Comment (juanmanuelramallo) [reply lib/action_view/component/base.rb]: One last question, what should happen if the file basename starts with a dot? I don't think we will ever encounter that scenario but better safe than sorry.

**Comment by juanmanuelramallo** [comment]
> @rdavid1099 @juanmanuelramallo any thoughts on how we'd like to proceed with this PR?

I think it's in a good state. 

We can go back to it later and add some tests following the other approach, my only comment was around the testing files.

**Comment by rdavid1099** [comment]
Thanks @joelhawksley @juanmanuelramallo. My schedule's getting packed with "end of year" chaos. Glad I could help out 😄
